### PR TITLE
1. feat(db): Store transactions in a separate database index, to improve query speed

### LINKED ISSUE #3151: Store transactions in a separate database index
## Motivation

`lightwalletd` caches all the blocks from Sapling activation, asking for them in height order. So the block RPC needs to be fast.

## Designs

https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures

For good performance, use the RocksDB `multi_get` methods to get all the transactions for a block.

## Tasks

1. Make the changes
2. Increment the database version
3. If you created new Rust modules, make sure the regenerate state paths are up to date in https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/test.yml

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Currently, Zebra reads the entire block each time we lookup a transaction. But we want to serve transactions to `lightwalletd` quickly and efficiently.

### Consensus Rules

This PR adds the genesis transaction hash to the transaction indexes, so Zebra can re-create the genesis block from its header and transaction.

There is a consensus rule that prevents spends of the genesis transparent UTXO. This PR doesn't change the UTXO index, so Zebra's consensus behaviour remains the same.

### Designs

The *Blocks* and *Transactions* parts of:
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures

## Solution

Database Changes:
- Implement disk serialization for block headers and transactions
- Add new transaction column families
- Split block header and transaction data
- Increment the database format to version 15

Related Database Changes:
- Re-write disk read API to avoid iterator-related hangs
- Move disk read API to ReadDisk
- Make the database generic over the thread mode, but don't actually change the thread mode in this PR

(In future PRs, we might need to change the thread mode for hang or performance testing. So I left the generic thread mode code in this PR.)

Testing:
- Update snapshot test data for transaction split
- Test that the new transaction indexes are consistent

To make sure there are no more hang bugs, I've started a full sync test on commit fce2ffe:
https://github.com/ZcashFoundation/zebra/actions/runs/2026544541

Closes #3151.

## Review

This PR is on the critical path for the lightwalletd work, so I've tagged it as a high priority.

 I tagged Conrado as the reviewer, because this PR will conflict with the `ZebraDb::transaction` changes in PR #3908. (It's only a minor conflict.)

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by conradoplg** [comment]
Rebased after base was merged

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (conradoplg) [inline zebra-state/src/service/finalized_state/disk_db.rs]: I'm kind of confused about this feature. Doesn't it require setting the `prefix_extractor`? e.g. https://github.com/rust-rocksdb/rust-rocksdb/issues/370#issuecomment-573399033

**Review by conradoplg — CHANGES_REQUESTED** [review]
Looks good, but I have a question about something that might need changing

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db.rs]: Thanks for catching this!

Prefix iteration requires setting a prefix extractor. But it also seems to cause database hangs, or tokio hangs, regardless of the database thread mode. So I've given up on prefix iteration.

Instead, I used a standard iterator. It's almost as efficient, because we can start at the start of the transactions, and use a Rust filter to stop iteration. See commit feaf30aff.

**Comment by teor2345** [comment]
I'm manually running a full sync test, because draft code in this PR had a hang bug:
https://github.com/ZcashFoundation/zebra/actions/runs/2024502648

**Comment by teor2345** [comment]
Turns out the current code still has a hang bug.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db.rs]: (And following commits, it turns out that any multi-key iteration takes a lock, which eventually causes a hang.)

**Comment by teor2345** [comment]
I tried multi-key iteration in this PR, but it seemed to cause hangs eventually. So I've gone back to single key fetches and manual iteration.

To make sure there are no more hang bugs, I've started a full sync test on commit fce2ffe:
https://github.com/ZcashFoundation/zebra/actions/runs/2026544541

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db.rs]: So this ended up not using iterators at all, except for very restricted single-key cases, wrapped in specific functions.

**Comment by teor2345** [comment]
> To make sure there are no more hang bugs, I've started a full sync test on commit [fce2ffe](https://github.com/ZcashFoundation/zebra/commit/fce2ffeae065f78f2306c5f17f79917dc8cf0541):
> https://github.com/ZcashFoundation/zebra/actions/runs/2026544541

Turns out that splitting transactions out might actually be faster than writing whole blocks.

(The full sync test passed in 5h46m, previous times were around 5h50m - 6h.)

**Review by conradoplg — DISMISSED** [review]
Looks good, I'll leave the unresolved conversation to be sure it's not merged yet

**Comment by teor2345** [comment]
> Looks good, I'll leave the unresolved conversation to be sure it's not merged yet

The `do-not-merge` label also stops merges, but only in mergify.

(I've added a conversation that explains why we're not merging yet.)

File: zebra-state/src/constants.rs
Comment (teor2345) [inline zebra-state/src/constants.rs]: Database format changes are risky changes, so we want to wait until after the testnet rollback release to merge this PR.

Comment (oxarbitrage) [reply zebra-state/src/constants.rs]: When is the rollback ?

File: zebra-state/src/constants.rs
Comment (oxarbitrage) [inline zebra-state/src/constants.rs:21]: I assume there is a PR bumping the database to `15` ?

File: zebra-state/src/constants.rs
Comment (teor2345) [reply zebra-state/src/constants.rs]: (At the engineering sync, we decided not to wait for the rollback, because it might take a few more days.)

File: zebra-state/src/constants.rs
Comment (teor2345) [reply zebra-state/src/constants.rs:21]: That was meant to be the NU5 PR #3799:
https://github.com/ZcashFoundation/zebra/pull/3799/files#diff-b5dc792d9507a26f7b72095ba80ace7587a23f285e92b8e2afdf4f438c5b2e02R21

But we decided to merge the database PRs first.

It seems like a lot of work to update 6 PRs to change one number.
Maybe we could just skip version 15? 🤔

Comment (teor2345) [reply zebra-state/src/constants.rs:21]: (I don't mind what we do here, happy to merge the PR and skip a version, someone else fixes the version, or I fix the version tomorrow.)

Comment (conradoplg) [reply zebra-state/src/constants.rs:21]: I think it's fine to skip 15

**Review by conradoplg — APPROVED** [review]
Re-approving

**Pull request merged**
